lib/pure/os.nim

```diff
@@ -3187,38 +3187,54 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
-proc sameFileContent*(path1, path2: string): bool {.rtl, extern: "nos$1",
-  tags: [ReadIOEffect], noWeirdTarget.} =
-  ## Returns true if both pathname arguments refer to files with identical
-  ## binary content.
+proc sameFileContent*(path1, path2: string; checkSize = false; checkFiles = false; bufferSize = -1): bool {.
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns `true` if both pathname arguments refer to files with identical binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ## This does not fail if the file never existed in the first place, unless `checkFiles = true`.
   ##
   ## See also:
   ## * `sameFile proc <#sameFile,string,string>`_
-  var
-    a, b: File
-  if not open(a, path1): return false
-  if not open(b, path2):
+  runnableExamples:
+    doAssert sameFileContent(currentSourcePath, currentSourcePath, checkSize = true, bufferSize = 4096)
+  var a, b: File
+  var mustRead = true
+  var bufferSize = if bufferSize == -1: getFileInfo(path1).blockSize else: bufferSize
```

Comment:
this would give `OSError` in case path1 is not stat-table, not honoring checkFiles.

the way it was written before you PR (from a File after it was already opened) avoids this, and avoids opening the file twice.

Reply:
The problem is that `alloc` that only takes `Natural` must be outside the `try` block and needs `bufferSize` too.

Reply:
easy:
```nim
var bufA, bufB: pointer
try:
  ...
  var bufferSize = if bufferSize == -1: ... else: ...
  bufA = alloc(bufferSize)
  ...
```

Reply:
Yeah I know that but I get `Attempt to read from nil` sometimes as you see on a few tests.